In [1]:
import pandas as pd
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

## 1. Understand Table Structure

*Checking the available columns and their data types*

In [4]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'return_data';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


column_name,data_type
value_return,double precision
qty_returned,integer
return_date,date
sku_id,character varying
nama_produk,character varying
status_return,character varying
return_reason,character varying
retur_id,character varying
invoice_id,character varying
so_id,character varying


**Findings**
- `value_return`typed as `float` → akan digunakan untuk menghitung total nilai kerugian akibat retur.
- `return_date` typed as `date` → kalkulasi tren bulanan dapat dilakukan langsung tanpa konversi.
- `nama_produk` and `sku_id` typed as `varchar` → `sku_id` akan digunakan sebagai foreign key untuk JOIN ke `master_sku` guna mendapatkan kategori produk.
- `return_reason` typed as `varchar` → akan digunakan untuk menganalisis alasan retur paling sering.
- `invoice_id` typed as `varchar` → akan digunakan sebagai foreign key untuk JOIN ke `invoice_data` guna mendapatkan informasi customer.

In [5]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'master_sku';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


column_name,data_type
price_max,double precision
price_min,double precision
kategori,character varying
sku_id,character varying
sku_status,character varying
satuan,character varying
sku_name,character varying


**Findings**
- `sku_id` typed as `varchar` →  primary key tabel ini, digunakan sebagai jembatan JOIN dari `return_data` ke `master_sku`
- `kategory`typed as `varchar` → akan digunakan untuk menganalisis kategory yang paling sering retur.
- `sku_name`typed as `varchar` → akan digunakan untuk menganalisis produk yang paling sering retur.

In [6]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'invoice_data';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
11 rows affected.


column_name,data_type
has_return,boolean
invoice_date,date
is_partial,boolean
partial_ke,double precision
invoice_value,double precision
bundle_complete_date,date
so_id,character varying
customer_id,character varying
invoice_id,character varying
payment_status,character varying


**Findings**
- `customer_id` typed as `varchar` →  primary key tabel ini, digunakan sebagai jembatan JOIN dari `return_data` ke `master_customer`
- `invoice_id` typed as `varchar` → akan digunakan sebagai foreign key untuk JOIN ke `return_data`

In [7]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'master_customer';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


column_name,data_type
customer_id,character varying
customer_name,character varying
customer_type,character varying
region,character varying
alamat,character varying
payment_terms,character varying
customer_status,character varying


**Findings**
- `customer_id` typed as `varchar` → primary key tabel ini, diakses melalui JOIN
- `region` typed as `varchar` → akan digunakan untuk menganalisis region yang paling sering retur.
- `customer_type` typed as `varchar` → akan digunakan untuk menganalisis tipe customer yang paling sering retur.

**Conclusion:**  
Semua tabel memiliki struktur yang valid. Foreign key antar tabel 
tersedia dan konsisten untuk mendukung JOIN di tahap analisis.

## 2. Data Preview

*Checking sample records to verify the data makes sense*

In [8]:
%%sql

SELECT *
FROM
    return_data
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


retur_id,invoice_id,so_id,sku_id,nama_produk,qty_returned,return_reason,return_date,status_return,value_return
RT00001,009502,SO240001,B006,Klem Arteri Bengkok,9,Salah Pesan Customer,2024-10-22,In Progress,1440000.0
RT00002,009510,SO240007,B014,Duk Operasi 80x90cm,22,Salah Pesan Customer,2024-07-21,Completed,1067000.0
RT00003,009515,SO240011,C019,NGT 16Fr,31,Kadaluarsa,2024-02-22,Completed,899000.0
RT00004,009541,SO240029,C006,Infus Set Anak,14,Salah Pesan Customer,2024-10-12,Completed,182000.0
RT00005,009543,SO240031,L010,Objek Glass Frosted,10,Salah Input Sistem,2024-08-22,Completed,855000.0


**Findings**
- Data terlihat valid dan masuk akal — `return_date` dalam format yang konsisten, `value_return` berisi nilai numerik yang wajar
- `return_reason` terlihat berisi nilai kategorikal — dari sample ini muncul Salah Pesan Customer, Kadaluarsa, dan Salah Input Sistem
- `status_return` memiliki dua nilai yang terlihat di sample ini: Completed dan In Progress


In [13]:
%%sql

SELECT *
FROM
    master_sku
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_id,sku_name,kategori,satuan,price_min,price_max,sku_status
C001,Masker Bedah 3-Ply,Consumable,Box,35000.0,55000.0,Active
C002,Sarung Tangan Nitril S,Consumable,Box,120000.0,180000.0,Active
C003,Sarung Tangan Nitril M,Consumable,Box,120000.0,180000.0,Active
C004,Sarung Tangan Nitril L,Consumable,Box,120000.0,180000.0,Active
C005,Infus Set Dewasa,Consumable,Pcs,8000.0,15000.0,Active


**Findings**
- `sku_name` terlihat valid — nama produk terbaca jelas dan konsisten
- `kategori` dari sample ini hanya terlihat 1 kategori (`consumable`), namun pengecekan nilai unik akan dilakukan di section selanjutnya


In [10]:
%%sql

SELECT *
FROM
    invoice_data
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


invoice_id,so_id,customer_id,invoice_date,is_partial,partial_ke,invoice_value,payment_status,status_invoice,bundle_complete_date,has_return
009723,SO240166,KLN068,2024-09-16,True,1.0,27138500.0,NET 30,Issued,None,False
010054,SO240404,KLN003,2024-02-12,True,1.0,6318000.0,NET 30,Bundle Complete,2024-03-07,False
009762,SO240195,RSS045,2024-02-23,True,1.0,21065000.0,NET 30,Bundle Complete,2024-03-20,False
009808,SO240226,RSP024,2024-11-18,True,2.0,14670000.0,NET 45,Issued,None,False
009926,SO240312,RSS066,2024-08-10,True,2.0,28354000.0,NET 30,Bundle Complete,2024-09-14,True


**Findings**
- `invoice_id` dan `customer_id` terlihat valid dan konsisten — siap digunakan sebagai jembatan JOIN dari `return_data` ke `master_customer`


In [11]:
%%sql

SELECT *
FROM
    master_customer
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


customer_id,customer_name,customer_type,region,alamat,payment_terms,customer_status
RSP001,RSUD Zulaika,RS Pemerintah,Jakarta Barat,"Jl. KH Amin Jasuta No. 173, Cirebon, JA 74749",NET 45,Active
RSS001,RS Suryono,RS Swasta,Tangerang,"Gang Dipatiukur No. 1, Tebingtinggi, Kepulauan Riau 46986",NET 30,Active
KLN001,Klinik Laksita,Klinik,Jakarta Timur,"Gang Surapati No. 0, Payakumbuh, Bali 83302",NET 30,Active
RSS002,RS Wasita,RS Swasta,Jakarta Utara,"Gg. Surapati No. 10, Surakarta, Maluku Utara 15272",NET 30,Active
KLN002,Klinik Sihombing,Klinik,Jakarta Selatan,"Jalan Rajawali Timur No. 088, Palangkaraya, BA 39649",NET 30,Active


**Findings**
- `region` terlihat valid — nama region terbaca jelas, namun pengecekan inkonsisten akan dilakukan di section selanjutnya
- `customer_type` terlihat berisi nilai kategorikal — dari sample ini muncul RS Pemerintah, RS Swasta, dan Klinik, namun pengecekan inkonsisten akan dilakukan di section selanjutnya


**Conclusion:**  
Data preview dari keempat tabel terlihat valid dan masuk akal. Potensi inkonsistensi penulisan ditemukan pada kolom `region` dan `customer_type` di `master_customer` — akan diverifikasi lebih lanjut di section Check Unique Values.

## 3. Count Total Records

*Checking the total number of rows in each table*

In [14]:
%%sql

SELECT COUNT(*) as total_records
FROM
    return_data;


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
59


In [15]:
%%sql

SELECT COUNT(*) as total_records
FROM
    master_sku;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
73


In [16]:
%%sql

SELECT COUNT(*) as total_records
FROM
    invoice_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
703


In [17]:
%%sql

SELECT COUNT(*) as total_records
FROM
    master_customer;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
220


**Findings**
- `return_data` memiliki 59 records — ini adalah total transaksi retur yang akan dianalisis sepanjang 2024
- `master_sku` memiliki 73 produk yang terdaftar
- `master_customer` memiliki 220 instansi kesehatan yang terdaftar
- `invoice_data` memiliki 703 records — jauh lebih banyak dari `return_data`, artinya tidak semua invoice berujung pada retur


**Conclusion:**  
Dataset cukup untuk dianalisis. Dari 703 invoice yang tercatat, hanya 59 yang menghasilkan retur — mengindikasikan retur tidak terjadi di semua transaksi.

## 4. Check Missing Values

*Checking for missing values in key columns*

In [22]:
%%sql

SELECT
    COUNT(*) - COUNT(value_return) AS missing_value_return,
    COUNT(*) - COUNT(return_date) AS missing_return_date,
    COUNT(*) - COUNT(nama_produk) AS missing_nama_product,
    COUNT(*) - COUNT(sku_id) AS missing_sku_id,
    COUNT(*) - COUNT(invoice_id) AS missing_invoice_id,
    COUNT(*) - COUNT(return_reason) AS missing_return_reason
FROM
    return_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_value_return,missing_return_date,missing_nama_product,missing_sku_id,missing_invoice_id,missing_return_reason
0,0,0,0,0,0


In [24]:
%%sql

SELECT
    COUNT(*) - COUNT(kategori) AS missing_kategory,
    COUNT(*) - COUNT(sku_name) AS missing_sku_name
FROM
    master_sku;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_kategory,missing_sku_name
0,0


In [27]:
%%sql

SELECT
    COUNT(*) - COUNT(customer_id) AS missing_customer_id
FROM
    invoice_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_customer_id
0


In [28]:
%%sql

SELECT
    COUNT(*) - COUNT(region) AS missing_region,
    COUNT(*) - COUNT(customer_name) AS missing_customer_name
FROM
    master_customer;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_region,missing_customer_name
0,0
